### DISCUSSION TITLE
an alternative pretty-printing format to help with constructing queries

### DISCUSSION THREAD

**Discussion by BrianHicks** [tree-sitter/tree-sitter#1675] [Libraries & tools]
for [tree-grepper](https://github.com/BrianHicks/tree-grepper), I've consistently wanted to have a pretty-printed view of the nodes in the tree and their content so I can create useful queries. While exploring tree-sitter-graph, I finally took a little time and made something nicer.

Things I wanted:

- Indent nodes to show hierarchy, as in `tree-sitter parse`
- Show just enough information for me to figure out where in a source file I should be looking

I came up with the following:

```
file 1:1
  module_declaration 1:1
    module 1:1: module
    upper_case_qid 1:8
      upper_case_identifier 1:8: Sample
    exposing_list 1:15
      exposing 1:15: exposing
      ( 1:24: (
      exposed_type 1:25
        upper_case_identifier 1:25: Test
        exposed_union_constructors 1:29
          ( 1:29: (
          double_dot 1:30: ..
          ) 1:32: )
      , 1:33: ,
      exposed_value 1:35
        lower_case_identifier 1:35: greeting
      ) 1:43: )
  type_declaration 4:1
    type 4:1: type
    upper_case_identifier 4:6: Test
    eq 5:5: =
    union_variant 5:7
      upper_case_identifier 5:7: Test
  value_declaration 8:1
    function_declaration_left 8:1
      lower_case_identifier 8:1: greeting
    eq 8:10: =
    string_constant_expr 9:5
      open_quote 9:5: "
      regular_string_part 9:6: Hello, World!
      close_quote 9:19: "
  value_declaration 12:1
    function_declaration_left 12:1
      lower_case_identifier 12:1: double
      lower_pattern 12:8
        lower_case_identifier 12:8: x
    eq 12:10: =
    bin_op_expr 13:5
      value_expr 13:5
        value_qid 13:5
          lower_case_identifier 13:5: x
      operator 13:7
        operator_identifier 13:7
          * 13:7: *
      number_constant_expr 13:9
        number_literal 13:9: 2
```

… which is the tree for this Elm file:

```elm
module Sample exposing (Test(..), greeting)

type Test
    = Test

greeting =
    "Hello, World!"

double x =
    x * 2
```

It's not 100% amazing—for example, I'd like to distinguish between named and anonymous nodes—but it's been sufficient to get me going with tree-sitter-graph.

The code, in case anyone else wants it, is at https://gist.github.com/BrianHicks/6afb43c0ed7ba731f73f46638cde0e69

I'd love to hear what other folks think of this. If it looks nice, I may publish my first-ever Rust crate. 😆

**Comment by ObserverOfTime** [comment]
The parse command has a `--cst` flag that outputs a similar format.
